satpy/modifiers/parallax.py

```diff
@@ -0,0 +1,474 @@
+                attribute describing the geolocation in a pyresample-aware way,
+                and they must contain satellite orbital parameters.  The
+                dimensions must be ``(y, x)``.  For best performance, this
+                should be a dask-based :class:`~xarray.DataArray`.
+
+        Returns:
+            :class:`~pyresample.geometry.SwathDefinition` describing parallax
+            corrected geolocation.
+        """
+        logger.debug("Calculating parallax correction using heights from "
+                     f"{cth_dataset.attrs.get('name', cth_dataset.name)!s}, "
+                     f"with base area {self.base_area.name!s}.")
+        (sat_lon, sat_lat, sat_alt_m) = _get_satpos_from_cth(cth_dataset)
+        self._check_overlap(cth_dataset)
+
+        cth_dataset = self._prepare_cth_dataset(cth_dataset)
+
+        (base_lon, base_lat) = self.base_area.get_lonlats(chunks=1024)
+        # calculate the shift/error due to the parallax effect
+        (shifted_lon, shifted_lat) = forward_parallax(
+                sat_lon, sat_lat, sat_alt_m,
+                base_lon, base_lat, cth_dataset.data)
+
+        shifted_area = self._get_swathdef_from_lon_lat(shifted_lon, shifted_lat)
+
+        # But we are not actually moving pixels, rather we want a
+        # coordinate transformation. With this transformation we approximately
+        # invert the pixel coordinate transformation, giving the lon and lat
+        # where we should retrieve a value for a given pixel.
+        (proj_lon, proj_lat) = self._get_corrected_lon_lat(
+                base_lon, base_lat, shifted_area)
+
+        return self._get_swathdef_from_lon_lat(proj_lon, proj_lat)
+
+    @staticmethod
+    def _get_swathdef_from_lon_lat(lon, lat):
+        """Return a SwathDefinition from lon/lat.
+
+        Turn ndarrays describing lon/lat into xarray with dimensions y, x, then
+        use these to create a :class:`~pyresample.geometry.SwathDefinition`.
+        """
+        # lons and lats passed to SwathDefinition must be data-arrays with
+        # dimensions,  see https://github.com/pytroll/satpy/issues/1434
+        # and https://github.com/pytroll/satpy/issues/1997
+        return SwathDefinition(
+            xr.DataArray(lon, dims=("y", "x")),
+            xr.DataArray(lat, dims=("y", "x")))
+
+    def _prepare_cth_dataset(self, cth_dataset):
+        """Prepare CTH dataset.
+
+        Set cloud top height to zero wherever lat/lon are valid but CTH is
+        undefined.  Then resample onto the base area.
+        """
+        # for calculating the parallax effect, set cth to 0 where it is
+        # undefined, unless pixels have no valid lat/lon
+        # NB: 0 may be below the surface... could be a problem for high
+        # resolution imagery in mountainous or high elevation terrain
+        # NB: how tolerant of xarray & dask is this?
+        cth_dataset = resample_dataset(
```

Comment:
This is reusing the variable name of the provided dataset. Can you use a different name for clarity and easier debugging.

Reply:
Fixed in 9509274e0f7735fc2a7a193e4623b923945f95a7.